tempesta_fw/t/unit/tfw_bomber.c

```diff
@@ -0,0 +1,496 @@
+}
+
+static void
+msg_send(tfw_bomber_desc_t *desc)
+{
+	struct sock *sk;
+	char *str;
+	int len, ret;
+	TfwStr msg;
+	TfwHttpMsg *req;
+	TfwMsgIter it;
+
+	sk = desc->sk;
+	BUG_ON(!sk);
+
+	atomic_inc(&tfw_bomber_request_nsend);
+
+	len = 1 * 1024 * 1024;
+	str = vmalloc(len);
+	if(!str) {
+		SS_ERR("Could not allocate memory for request\n");
+		return;
+	}
+	ret = fuzz_gen(str, str + len, 0, 1, FUZZ_REQ);
+	if (ret == FUZZ_END)
+		fuzz_reset();
+
+	msg.ptr = str;
+	msg.skb = NULL;
+	msg.len = strlen(str);
+	msg.flags = 0;
+
+	req = tfw_http_msg_create(&it, Conn_Clnt, msg.len);
+	tfw_http_msg_write(&it, req, &msg);
+	local_bh_disable();
+	ss_send(sk, &req->msg.skb_list, false);
+	local_bh_enable();
+
+	vfree(str);
+}
+
+static int
+tfw_bomber_connect_complete(struct sock *sk)
+{
+	int descidx, ret = 0;
+	tfw_bomber_desc_t *desc;
+	SsProto *proto = (SsProto *)sk->sk_user_data;
+
+	BUG_ON(proto == NULL);
+
+	descidx = proto->type;
+	desc = *(tfw_bomber_desc + descidx / tfw_connects)
+				 + descidx % tfw_connects;
+	BUG_ON(desc->proto.type != descidx);
+	BUG_ON(desc->proto.listener != NULL);
+	BUG_ON(desc->proto.hooks != &tfw_bomber_hooks);
+	BUG_ON(desc->sk && (desc->sk != sk));
+
+	desc->flags |= TFW_BOMBER_CONNECT_ESTABLISHED;
+	atomic_inc(&tfw_bomber_connect_ncomplete);
+	wake_up(&tfw_bomber_finish_wq);
+
+	return ret;
+}
+
+static int
+tfw_bomber_connection_close(struct sock *sk)
+{
+	int descidx;
+	tfw_bomber_desc_t *desc;
+	SsProto *proto = (SsProto *)sk->sk_user_data;
+
+	BUG_ON(proto == NULL);
+
+	descidx = proto->type;
+	desc = *(tfw_bomber_desc + descidx / tfw_connects)
+				 + descidx % tfw_connects;
+	BUG_ON(desc->proto.type != descidx);
+	BUG_ON(desc->proto.listener != NULL);
+	BUG_ON(desc->proto.hooks != &tfw_bomber_hooks);
+	BUG_ON(desc->sk && (desc->sk != sk));
+
+	desc->sk = NULL;
+	desc->flags |= TFW_BOMBER_CONNECT_CLOSED;
+	wake_up(&tfw_bomber_finish_wq);
+	return 0;
+}
+
+static int
+tfw_bomber_connection_error(struct sock *sk)
+{
+	int descidx;
+	tfw_bomber_desc_t *desc;
+	SsProto *proto = (SsProto *)sk->sk_user_data;
+
+	BUG_ON(proto == NULL);
+
+	descidx = proto->type;
+	desc = *(tfw_bomber_desc + descidx / tfw_connects)
+				 + descidx % tfw_connects;
+	BUG_ON(desc->proto.type != descidx);
+	BUG_ON(desc->proto.listener != NULL);
+	BUG_ON(desc->proto.hooks != &tfw_bomber_hooks);
+	BUG_ON(desc->sk && (desc->sk != sk));
+
+	desc->sk = NULL;
+	desc->flags |= TFW_BOMBER_CONNECT_ERROR;
+	atomic_inc(&tfw_bomber_connect_nerror);
+	wake_up(&tfw_bomber_finish_wq);
+	return 0;
+}
+
+static SsHooks tfw_bomber_hooks = {
+	.connection_new		= tfw_bomber_connect_complete,
+	.connection_drop	= tfw_bomber_connection_close,
+	.connection_error	= tfw_bomber_connection_error,
+};
+
+static void
+tfw_bomber_send_msgs(void)
+{
+	int i, k, m;
+
+	for (i = 0; i < tfw_threads; i++) {
+		for (k = 0; k < tfw_connects; k++) {
+			for (m = 0; m < tfw_messages; m++) {
+				if (tfw_bomber_desc[i][k].sk) {
+					msg_send(&tfw_bomber_desc[i][k]);
```

Comment:
Why do we do this in single thread? Rather we should send messages from _all_ the threads concurrently.